lib/api/WP_Auth0_Api_Refresh_Access_Token.php

```diff
@@ -0,0 +1,85 @@
+<?php
+/**
+ * Contains CUSTOM WP_Auth0_Api_Refresh_Access_Token.
+ *
+ * @package WP-Auth0
+ */
+
+/**
+ * Class WP_Auth0_Api_Refresh_Access_Token
+ * Get a new access token using the refresh token of a user.
+ *
+ * @see https://auth0.com/docs/tokens/refresh-token/current
+ */
+class WP_Auth0_Api_Refresh_Access_Token extends WP_Auth0_Api_Abstract {
+
+	/**
+	 * Default value to return on failure.
+	 */
+	const RETURN_ON_FAILURE = null;
+
+	/**
+	 * Make the API call and handle the response.
+	 *
+	 * @param string|null $code - Authorization code to exchange for tokens.
+	 * @param string|null $client_id - Client ID to use.
+	 * @param string|null $redirect_uri - Redirect URI to use.
+	 *
+	 * @return null|string
+	 */
+	public function call( $client_id = null, $client_secret = null, $refresh_token = null ) {
+		
+		if ( empty( $refresh_token ) ) {
+			return self::RETURN_ON_FAILURE;
+		}
+
+		$client_id = $client_id ?: $this->options->get( 'client_id' );
+		if ( empty( $client_id ) ) {
+			return self::RETURN_ON_FAILURE;
+		}
+
+		$client_secret = $client_secret ?: $this->options->get( 'client_secret' );
+		if ( empty( $client_secret ) ) {
+			return self::RETURN_ON_FAILURE;
+		}
+
+		return $this
+			->set_path( 'oauth/token' )
+			->add_body( 'grant_type', 'refresh_token' )
+			->add_body( 'client_id', $client_id )
+			->add_body( 'client_secret', $client_secret )
+			->add_body( 'refresh_token', $refresh_token )
+			->post()
+			->handle_response( __METHOD__ );
+
+	}
+
+	/**
+	 * Handle API response.
+	 *
+	 * @param string $method - Method that called the API.
+	 *
+	 * @return integer
+	 */
+	protected function handle_response( $method ) {
+
+		if ( 401 == $this->response_code ) {
+			WP_Auth0_ErrorManager::insert_auth0_error(
+				__METHOD__ . ' L:' . __LINE__,
+				__( 'An /oauth/token call triggered a 401 response from Auth0. ', 'wp-auth0' ) .
+				__( 'Please check the Client Secret saved in the Auth0 plugin settings. ', 'wp-auth0' )
+			);
+			return self::RETURN_ON_FAILURE;
+		}
+

```

Comment:
Let's also add a check for a `403` response, which would be an invalid refresh token or an Application that does not allow the refresh token grant (**Auth0 Dashboard > Applications > WP Application > Show Advanced Settings (bottom) > Grant Types**:

![Screenshot 2019-10-10 12 05 00](https://user-images.githubusercontent.com/855223/66598599-37f48f80-eb56-11e9-85e3-25eff468aef8.png)

The `error_description` that's returned by the API will work for the message used in the log.

Reply:
Hey Josh,
thanks for the suggestion :)!
Actually I thought this would already be covered by the call `$this->handle_failed_response( $method )`. I thought in case of a 403 the method of the `WP_Auth0_Api_Abstract` would call the `WP_Auth0_ErrorManager::insert_auth0_error` with the APIs error description. Doesn't it?
If not, I would implement a handling of the 403 close to the handling in this abstract method (if that's OK for you).

Reply:
@albeja - Good call, I think that should suffice.